test/model/test/common/wiredtiger_util.cpp

```diff
@@ -0,0 +1,262 @@
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    const char *value;
+    char cfg[64];
+
+    value = nullptr;
+
+    if (timestamp == 0)
+        testutil_check(session->begin_transaction(session, nullptr));
+    else {
+        testutil_snprintf(cfg, sizeof(cfg), "read_timestamp=%" PRIx64, timestamp);
+        testutil_check(session->begin_transaction(session, cfg));
+    }
+    testutil_check(session->open_cursor(session, uri, nullptr, nullptr, &cursor));
+
+    model::set_wt_cursor_key(cursor, key);
+    testutil_check_error_ok(ret = cursor->search(cursor), WT_NOTFOUND);
+    if (ret == 0)
+        testutil_check(cursor->get_value(cursor, &value));
+
+    testutil_check(cursor->close(cursor));
+    testutil_check(session->commit_transaction(session, nullptr));
+    return ret == 0 ? model::data_value(value) : model::NONE;
+}
+
+/*
+ * wt_insert --
+ *     Write to WiredTiger.
+ */
+int
+wt_insert(WT_SESSION *session, const char *uri, const model::data_value &key,
+  const model::data_value &value, model::timestamp_t timestamp, bool overwrite)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    char cfg[64];
+
+    testutil_check(session->begin_transaction(session, nullptr));
+    testutil_check(session->open_cursor(
+      session, uri, nullptr, overwrite ? nullptr : "overwrite=false", &cursor));
+
+    model::set_wt_cursor_key(cursor, key);
+    model::set_wt_cursor_value(cursor, value);
+    testutil_check_error_ok(ret = cursor->insert(cursor), WT_DUPLICATE_KEY);
+
+    testutil_check(cursor->close(cursor));
+    if (timestamp == 0)
+        cfg[0] = '\0';
+    else
+        testutil_snprintf(cfg, sizeof(cfg), "commit_timestamp=%" PRIx64, timestamp);
+    testutil_check(session->commit_transaction(session, cfg));
+
+    return ret;
+}
+
+/*
+ * wt_remove --
+ *     Delete from WiredTiger.
+ */
+int
+wt_remove(
+  WT_SESSION *session, const char *uri, const model::data_value &key, model::timestamp_t timestamp)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    char cfg[64];
+
+    testutil_check(session->begin_transaction(session, nullptr));
+    testutil_check(session->open_cursor(session, uri, nullptr, nullptr, &cursor));
+
+    model::set_wt_cursor_key(cursor, key);
+    testutil_check_error_ok(ret = cursor->remove(cursor), WT_NOTFOUND);
+
+    testutil_check(cursor->close(cursor));
+    if (timestamp == 0)
+        cfg[0] = '\0';
+    else
+        testutil_snprintf(cfg, sizeof(cfg), "commit_timestamp=%" PRIx64, timestamp);
+    testutil_check(session->commit_transaction(session, cfg));
+
+    return ret;
+}
+
+/*
+ * wt_update --
+ *     Update a key in WiredTiger.
+ */
+int
+wt_update(WT_SESSION *session, const char *uri, const model::data_value &key,
+  const model::data_value &value, model::timestamp_t timestamp, bool overwrite)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    char cfg[64];
+
+    testutil_check(session->begin_transaction(session, nullptr));
```

Comment:
There are two sets of `wt_` APIs, one set with and one without `txn`, which are not supposed to be mixed. The "non-txn" APIs do not expose transaction primitives to the caller, even though they have transactions inside them—but this is only so that we can use timestamps. The "txn" APIs expose transactions to the caller. And yes, as I am not aware of a mechanism to detect whether a session in the middle of a transaction, I'm just relying here on WT to return an error.